Reply/app/src/main/java/com/example/reply/ui/ReplyApp.kt

```diff
@@ -0,0 +1,286 @@
+                ReplyContentType.DUAL_PANE
+            } else {
+                ReplyContentType.SINGLE_PANE
+            }
+        }
+        WindowWidthSizeClass.Expanded -> {
+            navigationType = if (foldingDevicePosture is DevicePosture.BookPosture) {
+                ReplyNavigationType.NAVIGATION_RAIL
+            } else {
+                ReplyNavigationType.PERMANENT_NAVIGATION_DRAWER
+            }
+            contentType = ReplyContentType.DUAL_PANE
+        }
+        else -> {
+            navigationType = ReplyNavigationType.BOTTOM_NAVIGATION
+            contentType = ReplyContentType.SINGLE_PANE
+        }
+    }
+
+    /**
+     * Content inside Navigation Rail/Drawer can also be positioned at top, bottom or center for
+     * ergonomics and reachability depending upon the height of the device.
+     */
+    val navigationContentPosition = when (windowSize.heightSizeClass) {
+        WindowHeightSizeClass.Compact -> {
+            ReplyNavigationContentPosition.TOP
+        }
+        WindowHeightSizeClass.Medium,
+        WindowHeightSizeClass.Expanded -> {
+            ReplyNavigationContentPosition.CENTER
+        }
+        else -> {
+            ReplyNavigationContentPosition.TOP
+        }
+    }
+
+    ReplyNavigationWrapper(
+        navigationType = navigationType,
+        contentType = contentType,
+        navigationContentPosition = navigationContentPosition,
+        replyHomeUIState = replyHomeUIState,
+        closeDetailScreen = closeDetailScreen,
+        navigateToDetail = navigateToDetail
+    )
+}
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+private fun ReplyNavigationWrapper(
+    navigationType: ReplyNavigationType,
+    contentType: ReplyContentType,
+    navigationContentPosition: ReplyNavigationContentPosition,
+    replyHomeUIState: ReplyHomeUIState,
+    closeDetailScreen: () -> Unit,
+    navigateToDetail: (Long, ReplyContentType) -> Unit
+) {
+    val drawerState = rememberDrawerState(initialValue = DrawerValue.Closed)
+    val scope = rememberCoroutineScope()
+
+    val navController = rememberNavController()
+    val navigationActions = remember(navController) {
+        ReplyNavigationActions(navController)
+    }
+    val navBackStackEntry by navController.currentBackStackEntryAsState()
+    val selectedDestination =
+        navBackStackEntry?.destination?.route ?: ReplyRoute.INBOX
+
+    if (navigationType == ReplyNavigationType.PERMANENT_NAVIGATION_DRAWER) {
+        // TODO check on custom width of PermanentNavigationDrawer: b/232495216
+        PermanentNavigationDrawer(drawerContent = {
+            NavigationDrawerContent(
+                selectedDestination = selectedDestination,
+                navigationContentPosition = navigationContentPosition,
+                navigateToTopLevelDestination = navigationActions::navigateTo,
+            )
+        }) {
+            ReplyAppContent(
```

Comment:
This is slightly more advanced, but we should add in a `movableContentOf` usage here so that any and all state within `ReplyAppContent` is preserved even if the navigation type changes.

Reply:
I have never heard/used moveableContentOf, let me check it out :)

Reply:
Keeping it here for consideration and sharing.

```
val currentContentType by rememberUpdatedState(contentType)
val currentReplyHomeState by rememberUpdatedState(replyHomeUiState)

val rememberedReplyAppContent = remember {
    movableContentOf {
        ReplyNavHost(
            navController = navController,
            contentType = currentContentType,
            replyHomeUIState = currentReplyHomeState,
            navigationType = navigationType,
            closeDetailScreen = closeDetailScreen,
            navigateToDetail = navigateToDetail,
            modifier = Modifier.fillMaxSize(),
        )
    }
}

```

Reply:
Agreed to take a look at `movableContentOf` as a follow-up!